clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -298,35 +296,67 @@ protected void removeFromStorageEngine(int partition, byte[] keyBytes, Delete de
     }
   }
 
-  /** @return what kind of storage operation to execute, if any. */
-  private StorageOperationType getStorageOperationType(int partition, ByteBuffer valuePayload, ByteBuffer rmdPayload) {
-    PartitionConsumptionState pcs = partitionConsumptionStateMap.get(partition);
+  StorageOperationType getStorageOperationTypeForPut(int partition, Put put) {
+    PartitionConsumptionState pcs = getPartitionConsumptionStateMap().get(partition);
     if (pcs == null) {
       logStorageOperationWhileUnsubscribed(partition);
-      return StorageOperationType.NONE;
+      return StorageOperationType.SKIP;
     }
-    if (isDaVinciClient) {
+    ByteBuffer valuePayload = put.putValue;
+    ByteBuffer rmdPayload = put.replicationMetadataPayload;
+    checkStorageOperationCommonInvalidPattern(valuePayload, rmdPayload);
+
+    if (isDaVinciClient()) {
+      // For Da Vinci Client, RMD chunk should be ignored.
+      if (valuePayload.remaining() == 0) {
+        return StorageOperationType.SKIP;
+      }
       return StorageOperationType.VALUE;
     }
-    if (rmdPayload == null) {
-      throw new IllegalArgumentException("Replication metadata payload not found.");
+
+    if (rmdPayload.remaining() == 0) {
+      return StorageOperationType.VALUE;
     }
-    if (pcs.isEndOfPushReceived() || rmdPayload.remaining() > 0) {
-      // value payload == null means it is a DELETE request, while value payload size > 0 means it is a PUT request.
-      if (valuePayload == null || valuePayload.remaining() > 0) {
-        return StorageOperationType.VALUE_AND_RMD;
-      }
-      return StorageOperationType.RMD_CHUNK;
+    if (valuePayload.remaining() > 0) {
+      return StorageOperationType.VALUE_AND_RMD;
     } else {
+      // RMD chunk case.
+      return StorageOperationType.RMD_CHUNK;
+    }
+  }
+
+  StorageOperationType getStorageOperationTypeForDelete(int partition, Delete delete) {
+    PartitionConsumptionState pcs = getPartitionConsumptionStateMap().get(partition);
+    if (pcs == null) {
+      logStorageOperationWhileUnsubscribed(partition);
+      return StorageOperationType.SKIP;
+    }
+    ByteBuffer rmdPayload = delete.replicationMetadataPayload;
+    checkStorageOperationCommonInvalidPattern(null, rmdPayload);
+
+    if (isDaVinciClient()) {
+      if (pcs.isDeferredWrite()) {
+        return StorageOperationType.SKIP;
+      }
       return StorageOperationType.VALUE;
     }
+    return StorageOperationType.VALUE_AND_RMD;
```

Comment:
This doesn't look fully correct.
We didn't check whether rmd is empty or not, but just return `VALUE_AND_RMD` directly, and in theory, customer can run reprocessing job against AA store, which means the job can send delete without rmd before EOP, right?

There are quite a number of combinations, which can happen with different job types and before/after EOP, maybe it is good to call out all the cases we have considered.

Reply:
Delete without actually EOP can still use this operation type. The overhead is to put a dummy RMD in that specific key. 
I want to clarify that - with reprocessing job, this will only be before EOP and with deferred write set to false? I am wondering if we really want to add such case as it is covered by VALUE_AND_RMD case.